Add support for endpoints defined in Kestrel config

davidebbo opened the pull request:
Allow endpoints to be defined in the Kestrel config as an alternative to the profile. As is the case at runtime, Kestrel config endpoints override profile endpoints.

Fixes #3456

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/4205)

davidebbo committed:
Initial support for Kestrel endpoints

davidfowl commented on src/Aspire.Hosting/ProjectResourceBuilderExtensions.cs:
```diff
@@ -232,18 +232,42 @@ private static IResourceBuilder<ProjectResource> WithProjectDefaults(this IResou
+                if (endpointConfig.TryGetValue("Url", out var url) && url != null)
+                {
+                    // DISCUSS: check whether it's correct to replace * with localhost. Not sure what * really means, but it's not a valid in new Uri().
+                    url = url.Replace("*", "localhost");
```
[BindingAddress.Parse](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.bindingaddress.parse?view=aspnetcore-8.0)

davidebbo replied:
Ah, awesome!

davidebbo committed:
Use BindingAddress.Parse to parse the Url

davidebbo commented on playground/TestShop/BasketService/appsettings.json:
```diff
@@ -10,6 +10,9 @@
       "Protocols": "Http2"
+    },
+    "Endpoints": {
+      "Foo": { "Url": "http://*:5002" }
```
This probably wouldn't go in, as it's just to illustrate the PR. Though we'll need to write proper tests for the new feature.

davidebbo committed:
Simplify how we get the endpoint Url

davidfowl commented on src/Aspire.Hosting/ProjectResourceBuilderExtensions.cs:
```diff
@@ -232,18 +233,38 @@ private static IResourceBuilder<ProjectResource> WithProjectDefaults(this IResou
+            foreach (var endpoint in kestrelEndpoints)
             {
-                var uri = new Uri(url);
+                if (endpoint.GetValue<string>("Url") is var url && url != null)
```
```suggestion
                if (endpoint["Url"] is string url)
```

davidebbo committed:
Simplify getting url value

Co-authored-by: David Fowler <davidfowl@gmail.com>

davidebbo committed:
Fix Web Project detection in Publish mode

davidebbo committed:
- Refactor some Runtime vs Publish logic
- Use either scheme or name for binding name
- Set the transport even in Runtime mode

davidebbo committed:
Remove test change from appsettings.json

davidebbo committed:
Add initial Kestrel config test to discuss direction

davidfowl commented on src/Aspire.Hosting/IProjectMetadata.cs:
```diff
@@ -16,8 +17,9 @@ public interface IProjectMetadata : IResourceAnnotation
-    // This is for testing
+    // These are for testing
     internal LaunchSettings? LaunchSettings => null;
+    internal IConfiguration? KestrelConfiguration => null;
```
```suggestion
    internal IConfiguration? Configuration => null;
```

davidebbo replied:
Renamed throughout.

davidebbo committed:
rename KestrelConfiguration -> Configuration

davidfowl commented on src/Aspire.Hosting/ProjectResourceBuilderExtensions.cs:
```diff
@@ -223,27 +224,70 @@ private static IResourceBuilder<ProjectResource> WithProjectDefaults(this IResou
             builder.WithAnnotation(new LaunchProfileAnnotation(launchProfileName));
         }
 
+        var config = GetConfiguration(projectResource);
```
I think this code should run even if we've excluded the launch profile right?

davidebbo replied:
Ah yes, you're right, good catch. I'll rework this.

davidebbo commented on src/Aspire.Hosting/ProjectResourceBuilderExtensions.cs:
```diff
@@ -207,43 +208,84 @@ private static IResourceBuilder<ProjectResource> WithProjectDefaults(this IResou
         if (excludeLaunchProfile)
         {
             builder.WithAnnotation(new ExcludeLaunchProfileAnnotation());
-            return builder;
```
@davidfowl so if launch profile is excluded, we ignore any passed in `launchProfileName` (both before and after this change). It's slightly quirky, and throwing might be more correct, but that's a minor issue.

davidebbo committed:
Rework the logic when the profile is excluded

davidebbo committed:
Add more test coverage

davidebbo committed:
Extend Kestrel tests and move them to their own file

davidebbo committed:
Move ServiceC endpoint from profile to config

davidebbo committed:
Minor tweak

davidebbo committed:
Don't set ASPNETCORE_URLS when using Kestrel endpoints

davidebbo committed:
Remove TargetPort workaround now that issue is fixed

davidfowl commented:
Thanks @davidebbo !

fergalmoran commented:
So, as the person who originally raised this..
Where are we on getting this into bits we can play with?

davidebbo commented:
The change is in, but I don't know when it'll make it in some official build, @davidfowl are there daily builds people can get?

@fergalmoran or you could just build the repo and try it there, if only to verify that it works as you expect.

davidfowl commented:
Docs here on using the daily builds https://github.com/dotnet/aspire/blob/main/docs/using-latest-daily.md
